Adding k8s fields for Multi-App run for dev/test environment (#3719)
dapr/docs: daprdocs/content/en/developing-applications/local-development/multi-app-dapr-run/multi-app-overview.md

---

{{% alert title="Note" color="primary" %}}

>  Multi-App Run for **Kubernetes** is currently a preview feature only supported in Linux/MacOS.

Reviewer: @mukundansundar - Can you confirm this is true? Does it work on Windows.

Author: Was validated by @shivamkm07 that it works on windows

For a more in-depth example and explanation of the template properties, see [Multi-app template]({{< ref multi-app-template.md >}}).

{{% /codetab %}}

{{% codetab %}}
<!--kubernetes-->
With Multi-App Run, you can start multiple applications in a Kubernetes development or test environment using a single `dapr run -f -k` command using a template file. The template file describes how to start multiple applications as if you had run many separate CLI `run`commands. By default, this template file is called `dapr.yaml`. 

## Multi-App Run template file

> When you execute `dapr run -f .`, Dapr generates the multi-app template file (named `dapr.yaml`) in the `.dapr/deploy` directory within each application. Each time you run `dapr run -f -k .`, the generated Kubernetes Deployment YAML file is overwritten. run all the applications.

Reviewer: Is this a typo " run all the applications."

```suggestion
When you execute `dapr run -f .`, Dapr generates the multi-app template file (named `dapr.yaml`) in the `.dapr/deploy` directory within each application. Each time you run `dapr run -f -k .`, the generated Kubernetes Deployment YAML file is overwritten.
```

You can name template file with preferred name other than the default. For example `dapr run -f -k ./<your-preferred-file-name>.yaml`.

The following example includes some of the template properties you can customize for your applications. In the example, you can simultaneously launch 2 applications with app IDs of `nodeapp` and `pythonapp`.

```yaml
version: 1
common:
apps:
  - appID: nodeapp
    appDirPath: ./nodeapp/
    appPort: 3000
    containerImage: ghcr.io/dapr/samples/hello-k8s-node:latest
    createService: true
    env:
      APP_PORT: 3000
  - appID: pythonapp
    appDirPath: ./pythonapp/
    containerImage: ghcr.io/dapr/samples/hello-k8s-python:latest
```

> > **Note:** 

Reviewer: Should change this to a mark-down note format

Reviewer 2: The note/alert shortcode doesn't work inside {{< tabs >}}

Even if you've decided to rename your resources folder to something other than `.dapr`, the log files are written only to the `.dapr/logs` folder (created in the application directory).

## Watch the demo

Watch [this video for an overview on Multi-App Run](https://youtu.be/s1p9MNl4VGo?t=2456):

<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/s1p9MNl4VGo?start=2456" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

{{% /codetab %}}

{{% codetab %}}
<!--kubernetes-->

## Multi-App Run template file

When you execute `dapr run -k -f .` or `dapr run -k -f dapr.yaml`, the applications defined in the `dapr.yaml` Multi-App Run template file starts in Kubernetes default namespace. 

> > **Note:** Currently, the Multi-App Run template can only start applications in the default Kubernetes namespace.

Reviewer: @mukundansundar - Can you confirm this is true?

Author: Yes this is correct

Even if you've decided to rename your resources folder to something other than `.dapr`, the log files are written only to the `.dapr/logs` folder (created in the application directory).

## Watch the demo

Watch [this video for an overview on Multi-App Run](https://youtu.be/s1p9MNl4VGo?t=2456):

<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/s1p9MNl4VGo?start=2456" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

{{% /codetab %}}

{{% codetab %}}
<!--kubernetes-->

## Multi-App Run template file

> When you execute `dapr run -f -k .`, Dapr generates the multi-app template file (named `dapr.yaml`) in the `.dapr/deploy` directory within each application. Each time you run `dapr run -f -k .`, the generated Kubernetes Deployment YAML file is overwritten. 

Author: When the `dapr.yaml` template file is given as input (directory via `dapr run -k -f .`) or `dapr run -k -f dapr.yaml`, the necessary default service and deployment definitions for K8s are generated within the `.dapr/deploy` folder for each app in the `dapr.yaml` template. 
The `deployment.yaml` and if `createService` field is present in the `dapr.yaml` template then the `service.yaml` file(s) will be generated in the `.dapr/deploy` folder of each app in the multi-app run template and will be used to deploy the applications in a dev/test environment in K8s. 

For now the apps and services will be deployed in the `default` namespace of K8s.

Author: cc @dapr/maintainers-docs 
cc @hhunter-ms

Reviewer 2: @mukundansundar Updated - lmk what you think?

Even if you've decided to rename your resources folder to something other than `.dapr`, the log files are written only to the `.dapr/logs` folder (created in the application directory).

## Watch the demo

Watch [this video for an overview on Multi-App Run](https://youtu.be/s1p9MNl4VGo?t=2456):

<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/s1p9MNl4VGo?start=2456" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

{{% /codetab %}}

{{% codetab %}}
<!--kubernetes-->

## Multi-App Run template file

> When you execute `dapr run -k -f .` or `dapr run -k -f dapr.yaml`, the applications defined in the `dapr.yaml` multi-app run template file will start in Kubernetes default namespace. 

Reviewer: ```suggestion
When you execute `dapr run -k -f .` or `dapr run -k -f dapr.yaml`, the applications defined in the `dapr.yaml` multi-app run template file start in Kubernetes `default` namespace. 
```

Reviewer: Question: can you add have the applications in the another namespace? If so we should tell me people how. If not we should say this is a restriction.

Reviewer 2: per @mukundansundar - adding note about this

1. `appLogDestination` : This field configures the log destination for the application. The possible values are `console`, `file` and `fileAndConsole`. The default value is `fileAndConsole` where application logs are written to both console and to a file by default.

2. `daprdLogDestination` : This field configures the log destination for the `daprd` process. The possible values are `console`, `file` and `fileAndConsole`. The default value is `file` where the `daprd` logs are written to a file by default.

### Log file format

Logs for application and `daprd` are captured in separate files. These log files are created automatically under `.dapr/logs` directory under each application directory (`appDirPath` in the template). These log file names follow the pattern seen below:

- `<appID>_app_<timestamp>.log` (file name format for `app` log)
- `<appID>_daprd_<timestamp>.log` (file name format for `daprd` log)

Even if you've decided to rename your resources folder to something other than `.dapr`, the log files are written only to the `.dapr/logs` folder (created in the application directory).

> 

Reviewer: Can we add a link to the hello-Kubernetes tutorial https://github.com/dapr/quickstarts/tree/master/tutorials/hello-kubernetes for an example here and make this discoverable?

Even if you've decided to rename your resources folder to something other than `.dapr`, the log files are written only to the `.dapr/logs` folder (created in the application directory).

## Watch the demo

Watch [this video for an overview on Multi-App Run](https://youtu.be/s1p9MNl4VGo?t=2456):

<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/s1p9MNl4VGo?start=2456" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

{{% /codetab %}}

{{% codetab %}}
<!--kubernetes-->

## Multi-App Run template file

> Generate the multi-app template file (`dapr.yaml`) by running one of the following commands:

Author: This will not generate the multi-app run template file ... 
Based on the already existing `dapr.yaml` multi-app run template file, the the commands `dapr run -k -f .` or `dapr run -k -f dapr.yaml` will start the applications defined in the template in k8s `default` namespace.

```bash
dapr run -k -f .
```

Or

```bash
dapr run -k -f dapr.yaml
```

The necessary default service and deployment definitions for Kubernetes are generated within the `.dapr/deploy` folder for each app in the `dapr.yaml` template. 

> If the `createService` field is present in the `dapr.yaml` template and set to `true`, then the `service.yaml` file(s):
> - Generate in the `.dapr/deploy` folder of each app in the multi-app run template
> - Are used to deploy the applications in a dev/test environment in K8s 

Author: ```suggestion
If the `createService` field is set to `true` in the `dapr.yaml` template for an app, then the `service.yaml` file is generated in the `.dapr/deploy` folder of the app.
Otherwise only the `deployment.yaml` file is generated for each app that has the `containerImage` field set.
The files `service.yaml` and `deployment.yaml` are used to deploy the applications in `default` namespace in Kubernetes. This feature is specifically targeted only for running multiple apps in a dev/test environment in Kubernetes.
```
